Refactor structure of launch files reduce duplication when adding new simulations

glpuga opened the pull request:
- Refactor structure of launch files reduce duplication when adding new simulations
- Improve dockerfile to add ccache, avoid redundant image rebuilds and add build file persistance.

The gazebo simulation is now launched with:

```
ros2 launch ar4_gazebo_bringup main.launch.py
```


```mermaid
graph TD
      A[**ar4_moveit_config** <br> main.launch.py]
      B[**ar4_common** <br> main.launch.py]
      C[**ar4_description** <br> main.launch.py]
      D[**ar4_gazebo_sim** <br> main.launch.py]
      E[**ar4_gazebo_bringup** <br> main.launch.py]
      F[**ar4_hardware_interface** <br> main.launch.py]
      G[**ar4_realbot_bringup** <br> main.launch.py]

      A --> F
      C --> F

      H[**ar4_isaac_bringup** <br> main.launch.py <br> ...not yet available...]
      I[**...** <br> main.launch.py]

      B --> H
      I --> H

      J[**ar4_mujoco_bringup** <br> main.launch.py <br> ...not yet available...]
      K[**...** <br> main.launch.py]

      B --> J
      K --> J


      A --> B

      C --> D
      C --> A

      B --> E
      D --> E

      B --> G
      F --> G
```

glpuga committed:
Remove old ar4_gazebo package, will be split in multiple packages now

Signed-off-by: Gerardo Puga <glpuga@gmail.com>

glpuga committed:
Remove files to be replaced from robot description

Signed-off-by: Gerardo Puga <glpuga@gmail.com>

glpuga committed:
Refactor robot description to adapt for new structure

Signed-off-by: Gerardo Puga <glpuga@gmail.com>

glpuga committed:
Add new packages

Signed-off-by: Gerardo Puga <glpuga@gmail.com>

glpuga committed:
Refactor docker to avoid spurious rebuilds, add ccache, and external persistance

Signed-off-by: Gerardo Puga <glpuga@gmail.com>

glpuga committed:
Add bringup package for the real hardware

Signed-off-by: Gerardo Puga <glpuga@gmail.com>

glpuga committed:
Refactor README

Signed-off-by: Gerardo Puga <glpuga@gmail.com>

glpuga committed:
Fix pre-commit failures

Signed-off-by: Gerardo Puga <glpuga@gmail.com>

stevendes commented on ar4_common/README.md:
```diff
@@ -5,7 +5,7 @@ To build the package, run
 
 To run the simulation, source and run
-`ros2 launch ar4_gazebo ar4_sim.launch.py`
+`ros2 launch ar4_common ar4_sim.launch.py`
```
Is this correct? If that's the case, it is unclear what simulation we are launching here.

glpuga replied:
Not really. I did not see there was a readme in this folder. The documentation in the source folder should suffice.
I deleted it.
Good catch.

stevendes commented on ar4_common/package.xml:
```diff
@@ -0,0 +1,22 @@
+  <name>ar4_common</name>
+  <version>0.0.1</version>
+  <description>
+    Launch Gazebo simulation with Ar4 robot.
```
Is this description correct for the `ar4_common` package?

glpuga replied:
Fixed.

stevendes commented on ar4_gazebo_sim/LICENSE:
```diff
@@ -0,0 +1,29 @@
+BSD 3-Clause License
```
What advantage do we have in having `ar4_gazebo_sim` and `ar4_gazebo_bringup` as separate packages instead of just one `ar4_gazebo`?

glpuga replied:
Arguably, it would be totally possible have just one, so you can say this is mostly personal preference.

However, out of experience I find that keeping them separate allows to have a launch file that will only launch the simulation (or hardware) parts without the common parts, which comes handy when debugging the hardware/simulation layer.

Bringup packages end up being only a front-end for the packages needed for any particular configuration, and in the case of simulation, it's the single fail-safe place to set the use_sim_time param for the whole set without risking omissions.

Regarding naming, packages starting a robot (or a simulation of one) [must end in "bringup"](https://ros.org/reps/rep-0144.html#special-cases) to give a hint to the user of their function.

stevendes commented on ar4_realbot_bringup/launch/main.launch.py:
```diff
@@ -30,48 +28,48 @@
-        default_value='true',
-        description='Use simulation (Gazebo) clock if true',
-    )
+    """Launch the AR4 robot in Gazebo and MoveIt."""
```
This description talks about Gazebo but is under `ar4_realbot_bringup`, so I think this is an error.

glpuga replied:
Fixed.

stevendes commented on ar4_realbot_bringup/launch/main.launch.py:
```diff
@@ -30,48 +28,48 @@
-    return ld
+    return LaunchDescription(
+        [
+            use_sim_time_param,
```
Is `sim_time` true in the real robot?

glpuga replied:
Good catch, fixed. I started from a copy of the other package and this was left here by mistake.

stevendes reviewed (changes requested):
@glpuga I left some early comments, this change will make the repo incompatible with the [isaac_sim](https://github.com/Ekumen-OS/ar4/tree/isaac_sim) changes, just to keep that in mind, that branch has `AR4 + MoveIt` already added.

Is a test with the real robot necessary with these changes too?

glpuga committed:
Addressing review comments

glpuga committed:
fixup! Addressing review comments

glpuga committed:
fixup! fixup! Addressing review comments

glpuga committed:
fixup! fixup! fixup! Addressing review comments

glpuga committed:
fixup! fixup! fixup! fixup! Addressing review comments

glpuga commented:
> @glpuga I left some early comments, this change will make the repo incompatible with the [isaac_sim](https://github.com/Ekumen-OS/ar4/tree/isaac_sim) changes, just to keep that in mind, that branch has `AR4 + MoveIt` already added.

Indeed, I know. And potentially a Mujoco simulation coming (either by someone, or a test one I might try later today), but I think there's advantage both in reuse and clarity in the structure proposed here, and it'll get harder to refactor later.

Regarding the **isaac_sim** branch, looking at the changes, there should be no conflict with this branch in the robot description because I only added a single file in there, and changed no other.  The ar4_isaac package does not depend on anything I changed, so it can be merged with no conflict. We might have small conflicts on the README file and in the docker-compose file, though.

We can keep this PR on hold until we merge the isaac_sim branch, then I'll rebase this, fix conflicts and refactor the isaac_sim package in the same structure as this in this same PR.

We can then test also the hardware launch files before finally merging.

Thanks for the review.

glpuga commented:
Closing as this will be replaced with #29
